File: java/sort/array/CocktailSort.java
package sort.array;

public class CocktailSort {
    public void sort(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        while (start < end) {
            boolean flag = false;
            int index = 0;
            for (int i = start; i < end; ++i) {
                if (nums[i + 1] < nums[i]) {
                    swap(nums, i , i + 1);

                    flag = true;
                    index = i;
                }
            }

            if (!flag) {
                break;
            }
            end = index;

            flag = false;
            index = 0;
            for (int i = end; i > start; --i) {
                if (nums[i - 1] > nums[i]) {
                    swap(nums, i , i - 1);

                    flag = true;
                    index = i;
                }
            }

            if (!flag) {
                break;
            }
            start = index;
        }
    }

    private void swap(int[] nums, int i, int j) {
        nums[i] ^= nums[j];
        nums[j] ^= nums[i];
        nums[i] ^= nums[j];
    }
}

File: py/sort/array/CocktailSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class CocktailSort:
    def sort(self, nums: List[int]) -> None:
        start, end = 0, len(nums) - 1
        while start < end:
            flag, index = False, 0
            for i in range(start, end):
                if nums[i + 1] < nums[i]:
                    nums[i], nums[i + 1] = nums[i + 1], nums[i]
                    flag, index = True, i

            if not flag:
                break
            end = index

            flag, index = False, 0
            for i in range(end, start, -1):
                if nums[i - 1] > nums[i]:
                    nums[i], nums[i - 1] = nums[i - 1], nums[i]
                    flag, index = True, i

            if not flag:
                break
            start = index
